File: backend/src/models/media_user.py
from sqlalchemy.orm import relationship
from sqlalchemy import Column, Integer, ForeignKey, Date, SmallInteger

from backend.src.models.base import Base
from backend.src.models.statuses import Status
from backend.src.models.users import User


class MediaUser(Base):
    __tablename__ = 'media_user'

    id = Column(Integer, primary_key=True, index=True)
    date = Column(Date, nullable=True)
    rating = Column(SmallInteger, nullable=True)

    media_id = Column(Integer, ForeignKey('media.id'))
    status_id = Column(Integer, ForeignKey('statuses.id'))
    user_id = Column(Integer, ForeignKey('users.id'))

    media = relationship("Media", back_populates="users")
    status = relationship("Status")
    user = relationship("User", back_populates="media")

File: src/backend/src/main/java/thebandik/horizon/backend/user/mediaUser/MediaUser.java
package thebandik.horizon.backend.user.mediaUser;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;
import thebandik.horizon.backend.catalog.status.Status;
import thebandik.horizon.backend.media.Media;
import thebandik.horizon.backend.user.User;

import java.time.Instant;
import java.time.LocalDate;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(
        name = "media_user",
        uniqueConstraints = {
                @UniqueConstraint(columnNames = {
                        "media_id",
                        "user_id"
                })
        }
)
public class MediaUser {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "media_id", nullable = false)
    private Media media;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "status_id", nullable = false)
    private Status status;

    private Short rating;

    @Column(name = "first_event_date")
    private LocalDate firstEventDate;

    @Column(name = "last_event_date")
    private LocalDate lastEventDate;

    @Enumerated(EnumType.STRING)
    @JdbcTypeCode(SqlTypes.NAMED_ENUM)
    @Column(name = "last_event_precision")
    private DatePrecision lastEventPrecision;

    @Column(name = "history_count", nullable = false)
    private Integer historyCount = 0;

    @Column(name = "created_at", nullable = false)
    private Instant createdAt;

    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    @PrePersist
    public void onCreate() {
        createdAt = Instant.now();
        updatedAt = Instant.now();
    }

    @PreUpdate
    public void onUpdate() {
        updatedAt = Instant.now();
    }
}
